api/cluster/virtual_service.go

```diff
@@ -0,0 +1,188 @@
+	modelVersionHost, err := cfg.getModelVersionHost()
+	if err != nil {
+		return nil, err
+	}
+
+	// We manually create model version revision to avoid Kserve's modification to transformer url if the name contains "transformer"
+	// E.g. given a model version: std-transformer-s-157.namespace.caraml.dev
+	// The isvc.Status.URL would be: std-s-157-3-transformer.namespace.caraml.dev
+	// modelVersionRevisionHost, err := cfg.getModelVersionRevisionHost()
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	modelVersionRevisionHost := cfg.ModelVersionRevisionURL.Hostname()
+	modelVersionRevisionPath := cfg.ModelVersionRevisionURL.Path
+
+	vs := &v1beta1.VirtualService{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "networking.istio.io/v1beta1",
+			Kind:       "VirtualService",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cfg.Name,
+			Namespace: cfg.Namespace,
+			Labels:    cfg.Labels,
+		},
+		Spec: istiov1beta1.VirtualService{
+			Gateways: []string{knativeIngressGateway},
+			Hosts:    []string{modelVersionHost},
+			Http:     cfg.createHttpRoutes(modelVersionRevisionHost, modelVersionRevisionPath),
+		},
+	}
+
+	return vs, nil
+}
+
+// getModelVersionHost creates model version endpoint host based on version endpoint's url
+func (cfg *VirtualService) getModelVersionHost() (string, error) {
+	host := strings.Split(cfg.ModelVersionRevisionURL.Hostname(), fmt.Sprintf(".%s.", cfg.Namespace))
+	if len(host) != 2 {
+		return "", fmt.Errorf("invalid version endpoint url: %s. failed to split domain: %+v", cfg.ModelVersionRevisionURL, host)
+	}
+
+	domain := host[1]
+	return fmt.Sprintf("%s.%s.%s", cfg.Name, cfg.Namespace, domain), nil
+}
+
+// getModelVersionRevisionHost creates model version endpoint host based on version endpoint's url
+func (cfg *VirtualService) getModelVersionRevisionHost() (string, error) {
+	host := strings.Split(cfg.ModelVersionRevisionURL.Hostname(), fmt.Sprintf(".%s.", cfg.Namespace))
+	if len(host) != 2 {
+		return "", fmt.Errorf("invalid version endpoint url: %s. failed to split domain: %+v", cfg.ModelVersionRevisionURL, host)
+	}
+
+	domain := host[1]
+	return fmt.Sprintf("%s-%s.%s.%s", cfg.Name, cfg.RevisionID, cfg.Namespace, domain), nil
+}
+
+func (cfg *VirtualService) createHttpRoutes(modelVersionRevisionHost, modelVersionRevisionPath string) []*istiov1beta1.HTTPRoute {
+	routeDestinations := []*istiov1beta1.HTTPRouteDestination{
+		{
+			Destination: &istiov1beta1.Destination{
+				Host: defaultIstioIngressGatewayHost,
+			},
+			Headers: &istiov1beta1.Headers{
+				Request: &istiov1beta1.Headers_HeaderOperations{
+					Set: map[string]string{
+						"Host": modelVersionRevisionHost,
+					},
+				},
+			},
+			Weight: 100,
+		},
+	}
+
+	switch cfg.Protocol {
+	case protocol.UpiV1:
+		return []*istiov1beta1.HTTPRoute{
+			{
+				Route: routeDestinations,
+			},
+		}
+
+	default:
+		// TODO: Remove setting Content-Type header after all clients are updated to set the header
```

Comment:
IMO, we can keep it as is - we are going to create this new VS anyway, it can be the fallback handling if no content-type is set for HTTP protocol.

Reply:
Btw, can we only set the "Content-Type" header optionally, for those requests that are not setting this header? Reason is, if someone is sending binary data or so for some of their models, we shouldn't change the value to `"application/json"`. In general, we can ask the users to set the `Content-Type` header for all requests to GRPC models and especially to HTTP models that are not using `"application/json"`.

Reply:
Thanks again for reviewing the edge cases. As suggested, I created 3 route destinations:

For request targeting `/v1/models/{model_name}-{model_version}:predict` URI:
1. If content-type is empty, we will set content-type to application/json
2. If content-type is already set, we do nothing

3. And for other requests, we simply forward it to model version revision.

Reply:
The discussion is continuing [here](https://github.com/caraml-dev/merlin/pull/471#discussion_r1358057199), resolving this thread.